Add switch to package list command to show HTTP start and stop logs

joelverhagen opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Home/issues/9600
Regression: No

## Fix

This introduces a switch `--show-protocol-logs` to the `package list` command in NuGet.CommandLine.Xplat. For example:

```
> dotnet run --framework netcoreapp2.1 -- package list .\NuGet.CommandLine.XPlat.csproj --deprecated --verbosity normal

The following sources were used:
   https://api.nuget.org/v3/index.json
   C:\Users\joelv\.nuget\packages
   C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\

  GET https://api.nuget.org/v3/registration5-gz-semver2/microsoft.sourcelink.github/index.json
  GET https://api.nuget.org/v3/registration5-gz-semver2/microsoft.netcore.app/index.json
  GET https://api.nuget.org/v3/registration5-gz-semver2/system.runtime.serialization.primitives/index.json
  GET https://api.nuget.org/v3/registration5-gz-semver2/microsoft.extensions.commandlineutils/index.json
  GET https://api.nuget.org/v3/registration5-gz-semver2/microsoft.codeanalysis.fxcopanalyzers/index.json
  GET https://api.nuget.org/v3/registration5-gz-semver2/microsoft.build/index.json
  GET https://api.nuget.org/v3/registration5-gz-semver2/microsoft.build.locator/index.json
  OK https://api.nuget.org/v3/registration5-gz-semver2/microsoft.build/index.json 248ms
  OK https://api.nuget.org/v3/registration5-gz-semver2/microsoft.sourcelink.github/index.json 345ms
  OK https://api.nuget.org/v3/registration5-gz-semver2/microsoft.netcore.app/index.json 363ms
  OK https://api.nuget.org/v3/registration5-gz-semver2/microsoft.extensions.commandlineutils/index.json 1205ms
  OK https://api.nuget.org/v3/registration5-gz-semver2/microsoft.codeanalysis.fxcopanalyzers/index.json 1360ms
  OK https://api.nuget.org/v3/registration5-gz-semver2/system.runtime.serialization.primitives/index.json 3338ms
  OK https://api.nuget.org/v3/registration5-gz-semver2/microsoft.build.locator/index.json 3417ms
The given project `NuGet.CommandLine.XPlat` has no deprecated packages given the current sources
```

And without the switch (today's behavior):

```
> dotnet run --framework netcoreapp2.1 -- package list .\NuGet.CommandLine.XPlat.csproj --deprecated

The following sources were used:
   https://api.nuget.org/v3/index.json
   C:\Users\joelv\.nuget\packages
   C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\

The given project `NuGet.CommandLine.XPlat` has no deprecated packages given the current sources.
```
## Alternatives

I'm not totally sold on my own implementation, but it's the best of the options I considered. I considered these other options:

1. Adding a `-v|--verbosity <verbosity>` option. This seems like the best option but it poses two problems:
   1. The log level is [hard coded at the root of the application](https://github.com/NuGet/NuGet.Client/blob/eb9ad919724e5ef435d6321b775868c9f2c0f9a2/src/NuGet.Core/NuGet.CommandLine.XPlat/Program.cs#L76), so it would be strange to have a sub-command override it. In fact it would be tricky since log level is not mutable on `ILogger` which is what `list packages` receives.
   1. If the default log level (`Information`) is left, then HTTP logs would always be visible since they are [logged at the `Information` level](https://github.com/NuGet/NuGet.Client/blob/eb9ad919724e5ef435d6321b775868c9f2c0f9a2/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs#L111-L115). But I don't think we could safely change the default to `Minimal` (thus hiding HTTP logs by default) because that could affect other command's output in undesirable ways.
1. Adding a `--verbose` switch. If I recall correctly, this was removed from nuget.exe at some point in favor of `--verbosity`. Seems like it's the wrong word because of that.
1. Adding a `-d|--diagnostics` switch similar to `dotnet --diagnostics`. I chose not to go this route because I don't know how it would interact with the top level option and from [a GitHub issue](https://github.com/dotnet/sdk/issues/10309#issuecomment-499211751) it seems like the semantics are different.

~So I went with another switch that is more descriptive of what's actually happening. If we introduce `--verbosity` later, it could interact with this new switch however we want (e.g. automatically turn on protocol logs when verbosity is more verbose than `Information` or something like that).~

~I'm totally open to other ideas. I just wanted to put this PR out there to show that I'm willing to do the work. I'm happy to take another approach if it's better.~

I went with option 1 per the design spec.

## Testing/Validation

Tests Added: Yes
Validation: Manual testing with `dotnet run`

I would like to try making the changes to the [.NET SDK wrapper](https://github.com/dotnet/sdk/blob/master/src/Cli/dotnet/commands/dotnet-list/dotnet-list-package/ListPackageReferencesCommandParser.cs) but I don't know how to patch the bits locally. Any pointer in this direction would be great!

joelverhagen commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Program.cs:
```diff
@@ -193,7 +193,7 @@ private static CommandLineApplication InitializeApp(string[] args, CommandOutput
+                ListPackageCommand.Register(app, () => loggerFunc(log)(), () => new ListPackageCommandRunner());
             }
             else
             {
```
By the way, `loggerFunc` below 👇 is invoked outside of the delegate, so `DeleteCommand`, `PushCommand`, and `LocalsCommands` all have `HidePrefixForInfoAndMinimal = true`. From the shape of the code it doesn't look intentional but 🤷.

donnie-msft replied:
Are you sure? `Register` takes a Func<ILogger>, didn't think that would be an invocation.

joelverhagen replied:
On the previous line `CommandParsers.Register(app, loggerFunc(log));`, the `loggerFunc` is being invoked thus changing the state of `HidePrefixForInfoAndMinimal` but also returning a function.

joelverhagen commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/XPlatUtility.cs:
```diff
@@ -18,7 +18,6 @@ namespace NuGet.CommandLine.XPlat
     internal static class XPlatUtility
     {
         public const string HelpOption = "-h|--help";
-        public const string VerbosityOption = "-v|--verbosity <verbosity>";
```
Unused.

joelverhagen commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -204,10 +204,6 @@
   <data name="Source_Description" xml:space="preserve">
     <value>Package source (URL, UNC/folder path or package source name) to use. Defaults to DefaultPushSource if specified in NuGet.Config.</value>
   </data>
-  <data name="Switch_Verbosity" xml:space="preserve">
```
Unused.

joelverhagen commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:
```diff
@@ -611,7 +611,13 @@ private UpdateLevel GetUpdateLevel(NuGetVersion resolvedVersion, NuGetVersion la
             var sourceRepository = Repository.CreateSource(providers, packageSource, FeedType.Undefined);
```
Side note, shouldn't there just be one source repository created per source? It looks like a new one is created for every since ID. Not sure of the perf impact but that seems wrong...

nkolev92 replied:
I think that's incorrect. 
Just no one caught it.

joelverhagen replied:
I will leave this out of scope but it does bother me.

nkolev92 replied:
list package has a bug in it's parallelization too :D 

There are a few things that can be cleaned up at a later point.

nkolev92 commented:
fwiw, I'd go with the `--verbosity` switch. 
I think the `--show-protocol-logs` is a leaky abstraction. 

The default level for dotnet.exe commands is minimal and `dotnet nuget` should do the same thing.

nkolev92 commented:
cc @chgill-MSFT @anangaur 

@joelverhagen If you are motivated, you can whip up a design doc as per: https://github.com/NuGet/NuGet.Client/blob/dev-nkolev92-featureanddesignguide/docs/feature-guide.md

and we can discuss it :)

joelverhagen commented:
Sure, sounds good. I'll come up with a draft sometime.

joelverhagen commented:
I have added a spec: https://github.com/NuGet/Home/pull/9630

donnie-msft commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs:
```diff
@@ -85,10 +88,36 @@ public static class ListPackageCommand
                 {
                     var logger = getLogger();
 
+                    switch (verbosity.Value()?.ToUpperInvariant())
```
This feels weird. Do other dotnet tools embed verbosity strings? 
We do have NuGet.Common.LogLevel

joelverhagen replied:
AFAIK this is the first place where such logic exists. It's a mapping from MSBuild verbosity to NuGet log level, so only NuGet code could possibly have this. The existing places in the code have the reverse mapping, such as https://github.com/NuGet/NuGet.Client/blob/05925a63fde32ab277ccaab13ed466add9ac9dc8/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkLogger.cs.

Please let me know if you'd like to see this in a different shape or place.

joelverhagen replied:
I moved the code to a helper method per @nkolev92's feedback. I think it looks a little less out of place that way.

donnie-msft commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs:
```diff
@@ -435,6 +436,56 @@ public async Task DotnetListPackage_ProjectWithInitialTargets_Succeeds()
+                _fixture.RestoreProject(workingDirectory, projectName, string.Empty);
+
+                // Act
+                var listResult = _fixture.RunDotnet(
```
use explicit type

donnie-msft commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -4203,10 +4203,10 @@ public void PackCommand_PackageIcon_HappyPath_Warns_Succeeds()
-                .WithItem("None", "other\\files.txt", null)
-                .WithItem("None", "folder\\**", "media")
-                .WithItem("None", "utils\\*", "utils");
+                .WithItem("None", "icon.jpg", packagePath: string.Empty, pack: "true")
+                .WithItem("None", "other\\files.txt", packagePath: null, pack: "true")
+                .WithItem("None", "folder\\**", packagePath: "media", pack: "true")
+                .WithItem("None", "utils\\*", packagePath: "utils", pack: "true");
```
Should name all 4 params

joelverhagen replied:
Done.

donnie-msft commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -4261,7 +4261,7 @@ public void PackCommand_PackageIcon_PackWithSymbols_Succeeds(string symbolPackag
                 .WithProjectName("test")
                 .WithPackageIcon("icon.jpg")
-                .WithItem("None", "icon.jpg", "icon.jpg");
+                .WithItem("None", "icon.jpg", packagePath: "icon.jpg", pack: "true");
```
Should name all 4 params

joelverhagen replied:
Done.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs:
```diff
@@ -85,10 +88,36 @@ public static class ListPackageCommand
 
+                    switch (verbosity.Value()?.ToUpperInvariant())
+                    {
+                        case "Q":
```
only thing I'd suggest is extracting it to a separate utility, but maybe that's premature at this point given that there are no other usages.

joelverhagen replied:
Done.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs:
```diff
@@ -435,6 +436,56 @@ public async Task DotnetListPackage_ProjectWithInitialTargets_Succeeds()
             }
         }
 
+        [PlatformTheory(Platform.Windows, Skip = "Enabled after .NET CLI integration for https://github.com/NuGet/Home/issues/9600")]
```
I'd create the dotnet/sdk issue. 
Then create a NuGet side issue that basically says, once this PR has been inserted + SDK changes have been merged, re-enable these tests. 

Basically skipped tests should be pointing to an open issue.

joelverhagen replied:
I talked to William and he said it was fine to link to the NuGet issue on the CLI side instead of having a dedicated dotnet/sdk issue. Let me know if that's not okay.
For re-enabling the tests, I added: https://github.com/NuGet/Home/issues/9800.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs:
```diff
@@ -435,6 +436,56 @@ public async Task DotnetListPackage_ProjectWithInitialTargets_Succeeds()
+        [InlineData("", false)]
+        [InlineData("--verbosity minimal", false)]
+        [InlineData("--verbosity normal", true)]
+        public void DotnetListPackage_VerbositySwitchTogglesHttpVisibility(string args, bool showsHttp)
```
did you verify this test will actually pass? :D Is the plan to fix it later?

joelverhagen replied:
I am doing that now.

joelverhagen replied:
I had to fix some things up and put in some temporary hacks to point to a private CLI and private NuGet build, but it worked out.

nkolev92 replied:
YEp, have had to do the same thing. 

@martinrrm is gonna try to do the same with the `--prerelease` option. 

He might shoot you some questions.

nkolev92 reviewed (approved):
Looks great! Thanks!

Added a couple of comments.

nkolev92 commented:
@joelverhagen 

I kicked off the CI. 
Not sure how long it's been since you've rebased but that might be worth doing. 

You can kick off the CI yourself by removing and adding the label.

joelverhagen committed:
Add --verbosity option to list package command (#3411)

Fixes https://github.com/NuGet/Home/issues/9600
Implements design introduced by https://github.com/NuGet/Home/pull/9630
Needs integration in the .NET CLI (dotnet/sdk repo) to work end-to-end
Integration tests will be enabled later, tracked by https://github.com/NuGet/Home/issues/9800

donnie-msft reviewed (approved):
@dominoFire should review as well

nkolev92 commented:
There's no feature owner for list, it's an demand. 
Given how long this PR has been open, I think it's ok to merge.

nkolev92 commented:
@donnie-msft has a point about the ownership thing...depends which way you slice it. 
@dominoFire is marked as CLI owner, but search/list is on demand :D

dominoFire commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/ProjectFileBuilder.cs:
```diff
@@ -52,9 +52,14 @@ public void Build(MsbuildIntegrationTestFixture fixture, string path)
         /// Adds and intem inside a &lt;ItemGroup/&gt; node in the following form:
         /// <c>&lt;{itemType} Include="{itemPath}" [PackagePath="{packagePath}"] /&gt;</c>
```
xmldoc should be updated

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -712,4 +708,8 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
   </data>
-</root>
+  <data name="Verbosity_Description" xml:space="preserve">
+    <value>Set the verbosity level of the command. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic].</value>
```
According to the Microsoft Style Guide, being concise and consistent with the patterns is better for scannable content. See:
https://docs.microsoft.com/en-us/style-guide/scannable-content/#be-brief-be-bold-be-clear

Most of the help in flags in `dotnet.exe list package` have a third person writing style.

```suggestion
	    <value>Sets command verbosity level. Allowed values are: q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic].</value>
```

joelverhagen replied:
I chose to align with the string used in .NET CLI. I think consistency is preferable. Let me know and I can change it if you feel strongly.

KathleenDollard replied:
I agree that consistency is important here. @dominoFire I noted this to consider for a change across the CLI. I'm not sure about the first change though. Feels like we may lose clarity in that.

dominoFire commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/ListPackageTests.cs:
```diff
@@ -19,56 +21,91 @@ public class ListPackageTests
-                    () => logger,
-                    () => mockCommandRunner.Object);
+        [Fact]
+        public void BasicListPackageParsing_InteractiveTakesNoArguments()
```
Test method names should describe what's the test result, as described in the [NuGet coding guidelines](https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#unit-test-method-naming)

For example:

```suggestion
        public void BasicListPackageParsing_InteractiveTakesNoArguments_ThrowsException()
```

dominoFire reviewed (commented):
Some comments on naming and string resources.

joelverhagen committed:
Update test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/ListPackageTests.cs

Co-authored-by: Fernando Aguilar <feaguila@microsoft.com>

joelverhagen committed:
Update XML doc
